base/compiler/typelimits.jl

```diff
@@ -521,22 +524,38 @@ function tmerge(lattice::PartialsLattice, @nospecialize(typea), @nospecialize(ty
             return anyrefine ? PartialStruct(aty, fields) : aty
         end
     end
+    # Don't widen const here - external AbstractInterpreter might insert lattice
+    # layers between us and `ConstsLattice`.
+    isa(typea, PartialStruct) && (typea = widenconst(typea))
+    isa(typeb, PartialStruct) && (typeb = widenconst(typeb))
 
     # type-lattice for PartialOpaque wrapper
-    if isa(typea, PartialOpaque) && isa(typeb, PartialOpaque) && widenconst(typea) == widenconst(typeb)
-        if !(typea.source === typeb.source &&
-             typea.parent === typeb.parent)
-            return widenconst(typea)
+    apo = isa(typea, PartialOpaque)
+    bpo = isa(typeb, PartialOpaque)
+    if apo && bpo
+        aty = widenconst(typea)
+        bty = widenconst(typeb)
+        if aty == bty
+            if !(typea.source === typeb.source &&
+                typea.parent === typeb.parent)
+                return widenconst(typea)
+            end
+            return PartialOpaque(typea.typ, tmerge(typea.env, typeb.env),
+                typea.parent, typea.source)
         end
-        return PartialOpaque(typea.typ, tmerge(typea.env, typeb.env),
-            typea.parent, typea.source)
+        typea = aty
+        typeb = bty
+    elseif apo
+        typea = widenconst(typea)
+    elseif bpo
+        typeb = widenconst(typeb)
     end
 
-    # no special type-inference lattice, join the types
-    typea, typeb = widenconst(typea), widenconst(typeb)
-    @assert isa(typea, Type); @assert isa(typeb, Type)
+    return tmerge(widenlattice(lattice), typea, typeb)
+end
 
-    return tmerge(JLTypeLattice(), typea, typeb)
+function tmerge(lattice::ConstsLattice, @nospecialize(typea), @nospecialize(typeb))
+    return tmerge(widenlattice(lattice), widenconst(typea), widenconst(typeb))
```

Comment:
So we may want to implement `Const`-specific handling here for potential external `AbstractInterpreter` that uses `ConstsLattice` but not `PartialsLattice`?

Reply:
Yeah, I suppose.

Reply:
Actually, no, we're fine. The only thing you can do in tmerge of `ConstsLattice` that does not have `PartialsLattice` is to check for equality of the constants (which we don't handle here, but that's a duplicate of the issue discussed above, where we need to split out the tmerge path that checks for sub-lattice and equality).

Reply:
Ah, right. Maybe we want to leave a comment and the TODO for splitting out the fast path and general sub-lattice operations?